docs/control-structures.rst

```diff
@@ -271,6 +271,14 @@ Ranges are created using the ``range`` function. The following examples are vali
 
 ``STOP`` is a literal integer greater than zero. ``i`` begins as zero and increments by one until it is equal to ``STOP``.
 
+.. code-block:: python
+
+    for i in range(stop, bound=N):
+        ...
+
+``stop`` can be any integer greater than zero. ``N`` must be a compile-time constant. ``i`` begins as zero and increments by one until it is equal to ``stop``. If ``stop`` is larger than ``N``, execution will revert at runtime.
```

Comment:
should be _unsigned_ integer no? Otherwise you would need to have a guarantee at compile-time that the integer stays positive. Also, what about the case `range(x, bound=1)` and `x` is set to `2` previously as part of the function. Shouldn't that one already revert at compile-time (and not at runtime as stated here)?

Reply:
i guess i could clarify to say that execution will also revert if `stop` is less than 0

Reply:
as for this other case - that kind of value propagation is not something the compiler does right now. if it did, it would fall under the same class of checks where like `assert x < 1` would also be caught at compile time.

Reply:
Wait, it'll revert if `stop > N`?

Reply:
yep, that is the whole point of the bound. this is clear in the tests in the implementing PR -https://github.com/vyperlang/vyper/pull/3537/files#diff-c3f9a43de8dca93393504fef8b3c6bd8c77a6be969ee749f59b49537285e6392R30-R31

Reply:
> i guess i could clarify to say that execution will also revert if stop is less than 0

But would an integer work at all if previously ensured to be positive? Eg.:

```vyper
@external
def fn(x: int256):
    assert x > 0
    for i in range(x, bound=100):
        ...
```

Reply:
> as for this other case - that kind of value propagation is not something the compiler does right now. if it did, it would fall under the same class of checks where like `assert x < 1` would also be caught at compile time.

Do you think this is still worth mentioning that such a compile-time check is not done in that case?

Reply:
> yep, that is the whole point of the bound. this is clear in the tests in the implementing PR -https://github.com/vyperlang/vyper/pull/3537/files#diff-c3f9a43de8dca93393504fef8b3c6bd8c77a6be969ee749f59b49537285e6392R30-R31

Okay, I think the example should contain language showing that you have to make sure it doesn't revert, like:

```vyper
for i in range(min(x, N), bound=N):
   ...
```

Reply:
> Okay, I think the example should contain language showing that you have to make sure it doesn't revert, like:
> 
> ```vyper
> for i in range(min(x, N), bound=N):
>    ...
> ```

the following sentence clearly states though that execution will revert. the example should demonstrate the most basic use case, which is that you don't need `min()`.

Reply:
> Do you think this is still worth mentioning that such a compile-time check is not done in that case?

not really, as it is optimizer dependent. we could improve constant propagation later, but it might still not do the compile-time check if optimizations are turned off, for example. and in general i would personally not expect a compiler to do that check, although i might be pleasantly surprised if it did.

Reply:
> > Okay, I think the example should contain language showing that you have to make sure it doesn't revert, like:
> > ```vyper
> > for i in range(min(x, N), bound=N):
> >    ...
> > ```
> 
> the following sentence clearly states though that execution will revert. the example should demonstrate the most basic use case, which is that you don't need `min()`.

It would be really helpful to have an example in the docs of the best way to do a loop where it won't revert, since that will be a common use case

Reply:
hmm can add something of the sort. although `range(min(x, N), bound=N):` will still revert if x is negative.

Reply:
> hmm can add something of the sort. although `range(min(x, N), bound=N):` will still revert if x is negative.

Not worried about that use case, this is a good example when I want to iterate over a "chunk" of some large array, over a series of transactions bounded by some max iteration to ensure it doesn't get overly full and can't be emptied

Reply:
> > yep, that is the whole point of the bound. this is clear in the tests in the implementing PR -https://github.com/vyperlang/vyper/pull/3537/files#diff-c3f9a43de8dca93393504fef8b3c6bd8c77a6be969ee749f59b49537285e6392R30-R31
> 
> Okay, I think the example should contain language showing that you have to make sure it doesn't revert, like:
> 
> ```vyper
> for i in range(min(x, N), bound=N):
>    ...
> ```

Wouldn't the `bound=N` in this example be redundant? 

IMO, it would be clearer if we state that this is an alternative pattern to consider if the user does not want a runtime revert:
```
for i in range(min(x, N)):
    ...
```

Reply:
> do we want to support it?

I would think this would cause some additional compiler introspection which might be complicated to perform. I think the use cases that require it are rare enough not to optimize for it. the bound still works as intended